### ISSUE TITLE
Question about the code

### ISSUE THREAD

**Issue by MeWannaSleep** [jaymody/speculative-sampling#3]
In the paper, the author says speculative sampling is able to output the same distribution as the target model, while in this code,both using the  greedy search decoding method,how come did't yield the same results with the same distribution?

**Comment by jaymody** [comment]
Hi @MeWannaSleep. You're right, there's a mistake in the code due to the way I'm performing greedy sampling.

The rejection criteria + resampling method only recover the target distribution if we sample from the draft probability distribution, which we are not doing when we take argmax. We can see that in the proof, the term $\mathbb{P}(\tilde x = x)$ is equivalent to $p(x)$, which obviously assumes that we are drawing a sample from the distribution $p$ (the draft model):

![Screenshot 2023-08-17 at 2 37 53 PM](https://github.com/jaymody/speculative-sampling/assets/26451316/3597aba9-0534-476c-a08f-b6d4e24ed37b)

When we take the argmax, $p$ is no longer aligned with how we sampled from it ($p$ is the softmax of the raw logits, even though we are picking the highest probability next word with 100% probability).

The fix is to actually sample from the distribution. To get greedy sampling behaviour, we just rework the distribution so that the highest likelihood next token has 100% probability, which we can achieve with `temperature=0`.

Will submit a PR to fix this, thanks for pointing this out.

**Comment by MeWannaSleep** [comment]
Hi,@jaymody ,Thx for your reply, yet I still have two questions
1, 

> which obviously assumes that we are drawing a sample from the distribution p(the target model)

I would argue p is the draft model, because in the proof 
![image](https://github.com/jaymody/speculative-sampling/assets/135443303/41e38095-a20c-4832-ac52-6ed4d64a89f2)
2,To get the same distribution, we set temperature =0,which change the probability of the highest probability token to 1,and that results in the same decoding outcome as greedy search, so using speculative sampling greedy search is the only decoding method we are left to use?

**Comment by jaymody** [comment]
Ah, yes, I meant to say "which obviously assumes that we are drawing a sample from the distribution p (**the draft model**)". That was just a typo. Edited the post to correct it.

Regarding the second point, I'm not sure what you mean. The proof is just showing that sampling using the rejection criteria + resampling method is **equivalent** to just sampling directly from the target model. So you can use a non-zero temperature, you can apply top-p and top-k, it doesn't matter, it will be equivalent to regular autoregressive decoding. 

The reason why speculative and autoregressive sampling give different results when `temperature!=0` is just due to randomness in sampling from the distribution `np.random.choice` and randomness in the acceptance process `np.random.random() < min(1, q[i][j] / p[i][j])`. But, the underlying distribution being sampled from will always be the same (no matter the temperature etc ...). The reason why the results are the same when `temperature=0` is because we've reduced the distribution to be equal to 1 at a single point, so the result of `np.random.choice` will be deterministic, and the result of `np.random.random() < min(1, q[i][j] / p[i][j])` will always be True if the draft and target model agree on the same greedy token, else it will be False (nothing is left to chance).

That is to say, speculative sampling giving a different result than autoregressive is no different than running autoregressive sampling with a different seed and getting a different result.

**Comment by MeWannaSleep** [comment]
@jaymody ,thx for your detailed reply, another question is if i want to implement top p sampling method, assume p=3,and we get a draft token distribution ,p(x)=[a,a,a,b,c], there are five possible tokens, the first three get the top 3 probability, ,should we normalize the distribution to make it p(x)=[1/3,1/3,1/3,0,0]?
![image](https://github.com/jaymody/speculative-sampling/assets/135443303/34460e13-88e6-44ea-9992-8db1342c149b)
the above operation applies to the red circled distribution, am I right?
